spec.bs

```diff
@@ -507,22 +507,56 @@ To <dfn>validate and convert auction ad config</dfn> given an {{AuctionAdConfig}
 1. If |config|["{{AuctionAdConfig/sellerExperimentGroupId}}"] [=map/exists=]:
   1. Set |auctionConfig|'s [=auction config/seller experiment group id=] to
     |config|["{{AuctionAdConfig/sellerExperimentGroupId}}"].
-1. If |config|["{{AuctionAdConfig/perBuyerSignals}}"] [=map/exists=], [=map/for each=] |key| →
-  |value| of |config|["{{AuctionAdConfig/perBuyerSignals}}"]:
-  1. Let |buyer| be the result of [=parsing an origin=] with |key|. If |buyer| is failure,
-    [=exception/throw=] a {{TypeError}}.
-  1. Let |signalsString| be the result of [=serializing a JavaScript value to a JSON string=], given
-    |value|.
-  1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer signals=][|buyer|] to
-    |signalsString|.
-1. If |config|["{{AuctionAdConfig/perBuyerTimeouts}}"] [=map/exists=], [=map/for each=] |key| →
-  |value| of |config|["{{AuctionAdConfig/perBuyerTimeouts}}"]:
-  1. If |key| equals to "*", then set |auctionConfig|'s [=auction config/all buyers timeout=]
-    to min(|value|, 500) milliseconds, and [=iteration/continue=].
-  1. Let |buyer| the result of [=parsing an origin=] with |key|. If |buyer| is failure,
-    [=exception/throw=] a {{TypeError}}.
-  1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer timeouts=][|buyer|] to
-    min(|value|, 500) milliseconds.
+1. If |config|["{{AuctionAdConfig/perBuyerSignals}}"] [=map/exists=]:
+  1. If |config|["{{AuctionAdConfig/perBuyerSignals}}"] is a {{Promise}}:
+    1. Set |auctionConfig|'s [=auction config/per buyer signals=] to
+      |config|["{{AuctionAdConfig/perBuyerSignals}}"].
+    1. Increment |auctionConfig|'s [=auction config/pending promise count=].
+    1. [=Upon fulfillment=] of |auctionConfig|'s [=auction config/per buyer signals=] with |result|:
+      1. If |result| is not a [=record=] mapping from {{USVString}} to {{any}}, [=exception/throw=]
+        a {{TypeError}}.
+      1. Set |auctionConfig|'s [=auction config/per buyer signals=] to a new [=ordered map=] whose
+        [=map/keys=] are [=origins=] and whose [=map/values=] are [=strings=].
+      1. [=map/For each=] |key| → |value| of |result|:
+        1. Let |buyer| be the result of [=parsing an origin=] with |key|. If |buyer| is failure,
+          set |auctionConfig|'s [=auction config/per buyer signals=] to failure and [=iteration/break=].
+        1. Let |signalsString| be the result of [=serializing a JavaScript value to a JSON string=], given
+          |value|.
+        1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer signals=][|buyer|] to
+          |signalsString|.
+      1. Decrement |auctionConfig|'s [=auction config/pending promise count=].

```

Comment:
I think we never reach this line if any of the above ones throw errors. Maybe that's not observable, since the promise rejects and the user/developer never ever sees the side-effects of the in-parallel code hanging forever waiting for this decrement (at least I think that's what would happen if IIUC). But we should consider some sort of cancellation mechanism. Maybe we can file a bug to follow-up on this. Does this concern make sense to you or am I missing something?

Reply:
I reworked it.  PTAL.